### PR TITLE
aoblkdir: remove hole filling mechanism

### PR DESCRIPTION
**Pull request by soumyadeep2007**
Background:

(1) Block directory rowcount hole filling: The firstRowNum is allocated on the basis of gp_fast_sequence and is not always contiguous with the last minipage entry's (firstRowNum + rowCount) value. Before we insert the next minipage entry, we "fill" the rowCount of the previous one so that the range appears contiguous between successive minipage entries.

(2) Hole filling in action:
insert into foo values(1);
 -> inserts entry A with (firstRowNum, rowCount) = (1, 1)
insert into foo values(1);
 -> updates previous entry A to (firstRowNum, rowCount) = (1, 100)
    since firstRowNum = 101 for the next insert, we fill the previous entry
    to ensure that there are no holes in the rowCount range.

Motivation:

There is no apparent reason why this mechanism is necessary and it becomes a hindrance for future work to support unique indexes. That work depends on the continuity of a block directory entry's range to determine unique index lookups.

Changes:

(1) Removing this mechanism establishes the invariant: fetching physical rows in the continuous range of a block directory entry's first and last rownumbers will always be successful.

(2) We enforce this invariant with suitable ERRORs inside the fetch machinery. However, since hole filling will still exist in older versions, we do a AORelationVersion bump and ERROR out only for the current version and onwards.

Co-authored-by: Ashwin Agrawal <aashwin@vmware.com>

### CODE REVIEW

**Review by jimmyyih — APPROVED** [review]
So the proposed patch is to stop updating previous aoblkdir entries' row_count that seems to "fill in unused gaps" (e.g. an entry with 5 rows is changed to show 100 rows).  Yeah, I'm not sure why that was being done either.

File: src/include/catalog/pg_appendonly.h [hunk 1]
```diff
@@ -69,10 +69,11 @@ typedef enum AORelationVersion
 											 * were introduced, see MPP-7251 and MPP-7372. */
 	AORelationVersion_PG83 = 3,				/* Same as Aligned64bit, but numerics are stored
 											 * in the PostgreSQL 8.3 format. */
+	AORelationVersion_PG12 = 4,
```
Comment (bmdoil) [inline src/include/catalog/pg_appendonly.h:72]: Should we add a short comment here similar to the previous versions?

File: src/include/catalog/pg_appendonly.h [hunk 2]
```diff
@@ -69,10 +69,11 @@ typedef enum AORelationVersion
...
 	MaxAORelationVersion                    /* must always be last */
 } AORelationVersion;
 
-#define AORelationVersion_GetLatest() AORelationVersion_PG83
+#define AORelationVersion_GetLatest() AORelationVersion_PG12
```
Comment (bmdoil) [inline src/include/catalog/pg_appendonly.h:76]: Now that AO relations are not binary compatible 6->7 we will need to think carefully about how to work around this for the upcoming 6->7 upgrade work.

Comment (soumyadeep2007) [reply src/include/catalog/pg_appendonly.h:76]: Hmm they should be compatible in spite of the version bump, we will allow older versions even in 7X - its just that one would not be able to create unique indexes on such old tables (code to be added later in unique index PR).

Still, might be worth testing upgrade once with an AO table, to validate.

**Pull request merged**